File: backend/api/schemas/catalog.py
from ast import Nonlocal
from enum import Enum

from pydantic import BaseModel, Field


# --- Topics ---
class TopicCreate(BaseModel):
    name: str


class TopicRead(TopicCreate):
    id: int


# ------ Cats ---------


class CatCreate(BaseModel):
    name: str
    parent_id: int | None = None


class CatRead(CatCreate):
    id: int
    path: str | None = None


# ------- Authors ---------


class AuthorBase(BaseModel):
    name: str | None = None
    bio: str | None = None
    picture_url: str | None = None


class AuthorUpdate(AuthorBase):
    pass


class AuthorCreate(AuthorBase):
    name: str


class AuthorRead(BaseModel):
    id: int
    name: str


class AuthorInfo(AuthorRead, AuthorBase):
    pass


# ---------- Pubs ------------


class PubBase(BaseModel):
    name: str | None = None
    info: str | None = None
    picture_url: str | None = None


class PubUpdate(PubBase):
    pass


class PubCreate(PubUpdate):
    name: str


class PubRead(BaseModel):
    id: int
    name: str
    picture_url: str | None = None


class PubInfo(PubRead, PubBase):
    pass


# --- Books ---


class BookBase(BaseModel):
    title: str | None = None
    price: str | None = None
    isbn: str | None = None
    annotation: str | None = None
    page_count: int | None = None
    binding: str | None = None
    pub_id: int | None = None
    author_ids: list[int] | None = None
    topic_ids: list[int] | None = None


class BookUpdate(BookBase):
    pass


class BookCreate(BookBase):
    title: str
    price: float


class BookRead(BaseModel):
    id: int
    title: str
    price: float
    mean_rating: float
    reviews_count: int
    picture_url: str | None = None


class BookInfo(BookRead, BookBase):
    publisher: PubRead | None = None
    authors: list[AuthorRead] | None = None
    cat: CatRead | None = None
    topics: list[TopicRead] | None = None


# --- Filter ---


class Sortby(str, Enum):
    popularity = "popularity"
    rating = "rating"
    price = "price"


class BookFilter(BaseModel):
    cat_id: int | None = None
    topic_ids: list[int] | None = None
    min_price: float | None = Field(None, ge=0)
    max_price: float | None = Field(None, gt=0)
    bindings: list[str] | None = None
    is_new: bool | None = None
    min_rating: float | None = None
    author_ids: list[int] | None = None
    pub_ids: list[int] | None = None
    search: str | None = None
    limit: int = 10
    offset: int = 0
    sort_by: Sortby = Sortby.popularity
    reverse: bool = False

File: frontend/src/types/catalog.ts
// ==========================================
// 1. TOPICS (Темы)
// ==========================================
export class TopicCreate {
  name: string = "";

  constructor(init?: Partial<TopicCreate>) {
    Object.assign(this, init);
  }
}

export interface TopicRead extends TopicCreate {
  id: number;
}

// ==========================================
// 2. CATS (Категории / Жанры)
// ==========================================
export class CatCreate {
  name: string = "";
  parent_id: number | null = null;

  constructor(init?: Partial<CatCreate>) {
    Object.assign(this, init);
  }
}

export interface CatRead extends CatCreate {
  id: number;
  path: string | null;
}

// ==========================================
// 3. AUTHORS (Авторы)
// ==========================================

// Классы для форм создания и редактирования автора
export class AuthorCreate {
  name: string = ""; // Обязательное поле для создания
  bio: string | null = null;

  constructor(init?: Partial<AuthorCreate>) {
    Object.assign(this, init);
  }
}

export class AuthorUpdate {
  name: string | null = null;
  bio: string | null = null;

  constructor(init?: Partial<AuthorUpdate>) {
    Object.assign(this, init);
  }
}

// Интерфейсы только для чтения данных из API
export interface AuthorRead {
  id: number;
  name: string;
}

export interface AuthorInfo extends AuthorRead {
  bio: string | null;
  picture_url: string | null;
}

// ==========================================
// 4. PUBS (Издательства)
// ==========================================

export class PubCreate {
  name: string = ""; // Обязательное при создании
  info: string | null = null;

  constructor(init?: Partial<PubCreate>) {
    Object.assign(this, init);
  }
}

export class PubUpdate {
  name: string | null = null;
  info: string | null = null;

  constructor(init?: Partial<PubUpdate>) {
    Object.assign(this, init);
  }
}

export interface PubRead {
  id: number;
  name: string;
  picture_url: string | null;
}

export interface PubInfo extends PubRead {
  info: string | null;
}

export interface BindingRead {
  id: number;
  name: string;
}

// ==========================================
// 5. BOOKS (Книги)
// ==========================================

// Классы для админки (Создание и редактирование книг)
export class BookCreate {
  title: string = ""; // Обязательное
  price: number = 0; // Обязательное
  isbn: string | null = null;
  is_new: boolean = false;
  is_bestseller: boolean = false;
  annotation: string | null = null;
  page_count: number | null = null;
  binding_id: number | null = null;
  pub_id: number | null = null;
  author_ids: number[] | null = null;
  topic_ids: number[] | null = null;

  constructor(init?: Partial<BookCreate>) {
    Object.assign(this, init);
  }
}

export class BookUpdate {
  cat_id: number | null = null;
  title: string | null = null;
  price: number | null = null;
  isbn: string | null = null;
  is_new: boolean | null = null;
  is_bestseller: boolean | null = null;
  annotation: string | null = null;
  page_count: number | null = null;
  binding_id: number | null = null;
  pub_id: number | null = null;
  author_ids: number[] | null = null;
  topic_ids: number[] | null = null;

  constructor(init?: Partial<BookUpdate>) {
    Object.assign(this, init);
  }
}

// Интерфейсы только для чтения каталога
export interface BookRead {
  id: number;
  title: string;
  price: number;
  mean_rating: number;
  reviews_count: number; // не забудьте добавить, она была на вашем скриншоте ошибок
  picture_url: string | null;
  is_new: boolean;
  is_bestseller: boolean;
}

export interface BookInfo extends BookRead {
  isbn: string | null;
  page_count: number | null;
  annotation: string | null;
  binding: BindingRead | null;
  pub: PubRead | null;
  authors: AuthorRead[] | null;
  cat: CatRead | null;
  topics: TopicRead[] | null;
}

// ==========================================
// 6. FILTER (Фильтрация каталога)
// ==========================================
export enum SortBy {
  popularity = "popularity",
  rating = "rating",
  price = "price",
}

export class BookFilter {
  cat_id: number | null = null;
  topic_id: number[] | null = null;
  min_price: number | null = null;
  max_price: number | null = null;
  binding_id: number[] | null = null;
  is_new: boolean | null = null;
  is_bestseller: boolean | null = null;
  min_rating: number | null = null;
  author_id: number[] | null = null;
  pub_id: number[] | null = null;
  search: string = "";
  limit: number = 12;
  offset: number = 0;
  sort_by: SortBy = SortBy.popularity;
  reverse: boolean | null = null;

  constructor(init?: Partial<BookFilter>) {
    Object.assign(this, init);
  }
}
